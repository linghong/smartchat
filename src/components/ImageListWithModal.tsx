import { useState, FC, KeyboardEvent } from 'react'
import { BsTrash3 } from "react-icons/bs"

import ImageModal from '@/src/components/ImageModal'

interface ImageListWithModalProps {
  imageSrc: string[];
  handleImageDelete: (index: number) => void;
  isDeleteIconShow?: boolean;
}

const ImageListWithModal: FC<ImageListWithModalProps> = ({imageSrc, handleImageDelete, isDeleteIconShow = false}) => {

  const [modalOpen, setModalOpen] = useState<boolean>(false);
  const [selectedImage, setSelectedImage] = useState<string>('');

  const openModal = (imgSrc: string) => {
    setSelectedImage(imgSrc)
    setModalOpen(true)
  }

  const closeModal = () => {
    setModalOpen(false)
  }

  return (
    <>
     <div className="flex flex-row justify-start w-80vw pt-3 px-3">
      {
        imageSrc.map((imgSrc:string, i: number) => (
          <div key = {i} className="w-20 h-full mx-2">
            <div              
              className="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100" 
              onClick={()=>openModal(imgSrc)}
              onKeyDown={(e) => e.key==='Enter' && openModal(imgSrc)}
              tabIndex={0}
              role="button"
              aria-label={`Click to view larger image ${i + 1}`}
              title={`Uploaded image thumbnail ${i + 1}`}
            >
              <div 
                className={`bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer`} 
                style={{ backgroundImage: imageSrc ? `url(${imgSrc})` : 'none'}}                
              />
            </div>
           {
           isDeleteIconShow  &&  
            <BsTrash3 
              onClick={()=>handleImageDelete(i)}
              className="text-grey-600 cursor-pointer float-right"
              aria-label={`Delete image ${i + 1}`}
              tabIndex={0}
            />
          }
          </div>
        )
      )}
    </div>
    {modalOpen && <ImageModal src={selectedImage} onClose={closeModal} />} 
  </>    
)}

export default ImageListWithModal