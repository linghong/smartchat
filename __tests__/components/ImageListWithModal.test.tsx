import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import ImageListWithModal from '@/src/components/ImageListWithModal'

describe('ImageListWithModal', () => {
  const imageSrc = [
    {
      base64Image: 'data:image/png;base64,base64Imagedata1',
      mimeType: 'image/png',
      size: 5000,
      name: 'image1',
    },
    {
      base64Image: 'data:image/png;base64,base64Imagedata2',
      mimeType: 'image/png',
      size: 8000,
      name: 'image2',
    },
  ]

  const handleImageDelete = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should render the list of images', () => {
    render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
      />,
    )

    const images = screen.getAllByRole('button', {
      name: /click to view larger image/i,
    })
    expect(images).toHaveLength(imageSrc.length)
  })

  it('should open the modal when an image is clicked', () => {
    render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
      />,
    )

    const image = screen.getByRole('button', {
      name: /click to view larger image 1/i,
    })
    fireEvent.click(image)

    const modal = screen.getByRole('dialog')
    expect(modal).toBeInTheDocument()

    const modalImage = screen.getByAltText('Expanded view')
    expect(modalImage).toBeInTheDocument()
    expect(modalImage).toHaveAttribute('src', imageSrc[0].base64Image)
  })

  it('should open the modal when Enter key is pressed', () => {
    render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
      />,
    )

    const image = screen.getByRole('button', {
      name: /click to view larger image 1/i,
    })
    fireEvent.keyDown(image, { key: 'Enter', code: 'Enter' })

    const modal = screen.getByRole('dialog')
    expect(modal).toBeInTheDocument()

    const modalImage = screen.getByAltText('Expanded view')
    expect(modalImage).toBeInTheDocument()
    expect(modalImage).toHaveAttribute('src', imageSrc[0].base64Image)
  })

  it('should close the modal when the close button is clicked', () => {
    render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
      />,
    )

    const image = screen.getByRole('button', {
      name: /click to view larger image 1/i,
    })
    fireEvent.click(image)

    const closeButton = screen.getByRole('button', {
      name: /close image view/i,
    })
    fireEvent.click(closeButton)

    expect(screen.queryByRole('dialog')).not.toBeInTheDocument()
  })

  it('should display the delete icon when isDeleteIconShow is true', () => {
    render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
        isDeleteIconShow={true}
      />,
    )

    const deleteIcons = screen.getAllByLabelText(/delete image/i)
    expect(deleteIcons).toHaveLength(imageSrc.length)
  })

  it('should not display the delete icon when isDeleteIconShow is false', () => {
    render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
        isDeleteIconShow={false}
      />,
    )

    const deleteIcons = screen.queryAllByLabelText(/delete image/i)
    expect(deleteIcons).toHaveLength(0)
  })

  it('should call handleImageDelete when the delete icon is clicked', () => {
    render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
        isDeleteIconShow={true}
      />,
    )

    const deleteIcon = screen.getByLabelText('Delete image 1')
    fireEvent.click(deleteIcon)

    expect(handleImageDelete).toHaveBeenCalledTimes(1)
    expect(handleImageDelete).toHaveBeenCalledWith(0)
  })

  it('should match snapshot', () => {
    const { asFragment } = render(
      <ImageListWithModal
        imageSrc={imageSrc}
        handleImageDelete={handleImageDelete}
      />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
