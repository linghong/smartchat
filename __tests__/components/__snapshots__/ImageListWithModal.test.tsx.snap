// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageListWithModal should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row justify-start w-80vw pt-3 px-3"
  >
    <div
      class="w-20 h-full mx-2"
    >
      <div
        aria-label="Click to view larger image 1"
        class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
        role="button"
        tabindex="0"
        title="Uploaded image thumbnail 1"
      >
        <div
          class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
          style="background-image: url(http://example.com/image1.jpg);"
        />
      </div>
    </div>
    <div
      class="w-20 h-full mx-2"
    >
      <div
        aria-label="Click to view larger image 2"
        class="mr-2 p-1 bg-clip-border bg-slate-100 border-2 border-violet-100"
        role="button"
        tabindex="0"
        title="Uploaded image thumbnail 2"
      >
        <div
          class="bg-no-repeat bg-center bg-contain min-h-[60px] h-full w-full cursor-pointer"
          style="background-image: url(http://example.com/image2.jpg);"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
